kube-runtime/src/reflector/object_ref.rs

```diff
@@ -2,14 +2,102 @@ use derivative::Derivative;
 use k8s_openapi::{api::core::v1::ObjectReference, apimachinery::pkg::apis::meta::v1::OwnerReference};
 use kube_client::{
     api::{DynamicObject, Resource},
-    core::ObjectMeta,
-    ResourceExt,
+    core::{api_version_from_group_version, ObjectMeta},
 };
 use std::{
+    borrow::Cow,
     fmt::{Debug, Display},
     hash::Hash,
 };
 
+/// An object that can be indexed by a [reflector store](super::Store).
+///
+/// This trait is blanket-implemented for all [`Resource`] objects.
+#[allow(clippy::module_name_repetitions)]
+pub trait ToObjectRef {
```

Comment:
To be fair I think the name Resource` was confusing to begin with. `Resource` is supposed to represent the Kubernetes object, but an object that only contains the metadata is barely a Kubernetes object anymore. Meanwhile, it is even more confusing to call it a `ResourceRef`, which gives the impression that it is something like an `AsRef<Resource>`.

Reply:
> it is even more confusing to call it a ResourceRef, which gives the impression that it is something like an AsRef<Resource>.

that is fair. but it also feels weird to call the resource `ToObjectRef` which makes it sound like something analogous to a `From` impl, but in reality it does everything `Resource` does, just more efficient

Reply:
> Resource is supposed to represent the Kubernetes object, but an object that only contains the metadata is barely a Kubernetes object anymore

not sure I agree with this. the generic parts of a resource can only be `metadata` + group/version/kind. it's strictly more than `Metadata` (because of GVK), and strictly less than the whole thing (because we cannot generically get arbitrary other fields). the name does mirror `k8s-openapi::Resource` so i feel naming it anything else runs the risk of more confusion.

as a - perhaps offensive, but not unserious - alternative suggestion, what about `ResourceCow`?

Reply:
@clux how is it a cow? It is just the minimal trait containing necessary methods to identify an object for indexing purposes.

Reply:
I think it was a bit fallacious of me to think of traits as the contained fields indeed. After all, implementors are almost always whole objects.

This is something like the utility vs semantic classes debate. From the perspective of utility, we actually want `Resource` to represent a generic whole object regardless (though that's actually weird to fit with PartialObjectMeta) and `ReflectorValue` to represent a trait that can be used in a reflector, while in terms of semantics, we want `ToMeta` and `ToObjectRef`, the former of which is apparently confusing for the user indeed.

Reply:
> This is something like the utility vs semantic classes debate.

indeed. normally we go more towards the semantic side, but naming things is always hard 🙃
definitely happy to take other suggestions.  i am just trying to avoid the normal landmines personally (no From/Into, and `verb-able` suffix).

Full semantic here is probably something like `ReflectorLookup`? Then we `impl ReflectorLookup for Resource` and is the constraint makes sense in the codebase, and anyone implementing it understands why they are implementing it from the trait name.

> how is it a cow?

well, the trait does basically what `Resource` does but "mostly return `Cow` instead of owned strings". to me this trait is very similar to `Resource` so makes sense if the name reflected that (but that was just a thought - and i like it less now).

Reply:
So what about `reflector::Entry`? `Lookup` gives me the impression that it is a key rather than a value.

Reply:
hm, it's not really a value either, it's behaviour that allows for resource lookup (and semantically we are looking for a verb). `Entry` has baggage and it implies it's some kind of [view into a collection](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html).

the trait name here does not need to be one-word short, it's possibly better for it to be informative, since users are not expected to use this most of the time.

Reply:
Yep, I just didn't want to [duplicate the module name](https://rust-lang.github.io/rust-clippy/master/index.html#/module_name_repetitions).

Reply:
How is it not a value stored inside a reflector though? Why would `reflector::Value` be inappropriate?

Reply:
A `reflector::Value` would be something you got out of Store itself, whereas this is the behaviour that allows you to get multiple different values from the real value (the object in the store)

Reply:
Maybe `StoreResource` ? `impl StoreResource for Resource`
it would possibly duplicate module name, but i also don't want to have two traits called `Resource` from kube

Reply:
how is this getting multiple different values from the store object?

Reply:
the trait has multiple methods, all of which grabs/returns values from the underlying store object

Reply:
Oh, by "values" you are referring to different fields. But aren't the implementors of the trait just the values in the store itself? When we get something out of the store, we are actually getting an `impl reflector::Value`.

Reply:
The signature of `get` is `pub fn get(&self, key: &ObjectRef<K>) -> Option<Arc<K>>`, not `impl Value`; the **users decide what `K` is** when they create the store. This PR removes `K: Resource`, but you still put a representation of kubernetes resources in there. We only minimised the trait requirements - allowing users to elide the whole `Resource` for a more minimal lookup trait that __acts on__ one.

Users will make a store for `Pod` or `Deployment` in the general case, and those types will implement this proposed trait so so that the signature of the store impls can be as light as possible. Users does not in-general need to know what a store Value/Entry is because they've already decided the type.

I think it's simplest if we do not try to confuse the users with new values, nor resource variant names (as i originally proposed), but describe the behaviour in the trait name (ala `StoreLookup` or `ReflectorLookup`) because;

1. the user is generally not going to see this trait
2. it will help runtime devs (oh yeah, it's the trait that reflector uses for lookups) - it makes sense semantically
3. if the users sees it they likely only encounter it incidentally through docs.rs for [reflector](https://docs.rs/kube/latest/kube/runtime/reflector/fn.reflector.html) or [Store](https://docs.rs/kube/latest/kube/runtime/reflector/struct.Store.html), where they are more likely to look at examples anyway
4. if advanced users use stores for really low level stuff (where they can't use our full `Resource` trait), that's the time we expect them to learn about the trait (perhaps via a "this struct does not satisfy trait condition `impl StoreLookup for K` required by `Store<K>`" error)